File: 43-multiply-strings.cpp
string multiply(string num1, string num2) {
    int m = num1.size();
    int n = num2.size();
    string sum(m+n, '0');

    for (int i = m-1; i >= 0; i--) {
        int carry = 0;
        for (int j = n-1; j >=0; j--) {
            carry += (sum[i+j+1] - '0') + (num1[i] - '0') * (num2[j] - '0');
            sum[i+j+1] = carry % 10 + '0';
            carry /= 10;
        }
        sum[i] += carry;
    }

    for (int i = 0; i < m + n; i++) {
        if (ans[i] != '0') {
            return ans.substr(i);
        }
    }

    return "0";
}

File: LeetCode/p0043/II/43-multiply-strings.java
// 标签：字符串、相乘

class Solution {
    public String multiply(String num1, String num2) {
        int m = num1.length();
        int n = num2.length();

        int[] pos = new int[m+n];

        for (int i = m - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                int product = (num1.charAt(i) - '0') * (num2.charAt(j) - '0');

                int pos1 = i + j;
                int pos2 = i + j + 1;

                int sum = product + pos[pos2];

                pos[pos1] += sum / 10;
                pos[pos2] = sum % 10;
            }
        }

        String result = "";

        int idx = 0;

        while (idx < m+n && pos[idx] == 0) {
            // 找到第一个不为0的位置
            idx += 1;
        }

        if (idx == m+n) {
            return "0";
        } else {
            for (int i = idx; i < m + n; i++) {
                result += pos[i];
            }
        }

        return result;
    }
}
